lib/shared/puppet_forge/tar/mini.rb

```diff
@@ -0,0 +1,74 @@
+require 'zlib'
+require 'archive/tar/minitar'
+
+module PuppetForge
+  class Tar
+    class Mini
+      def unpack(sourcefile, destdir, logger)
+        @logger = logger
+        dirlist = []
+        Zlib::GzipReader.open(sourcefile) do |reader|
+          Archive::Tar::Minitar.unpack(reader, destdir, find_valid_files(reader)) do |action, name, stats|
+            case action
+            when :file_done
+              FileUtils.chmod('u+rw,g+r,a-st', "#{destdir}/#{name}")
+            when :file_start
+              validate_entry(destdir, name)
+              @logger.debug("Extracting: #{destdir}/#{name}")
+            when :dir
+              validate_entry(destdir, name)
+              dirlist << "#{destdir}/#{name}"
+              @logger.debug("Extracting: #{destdir}/#{name}")
+            end
+          end
+        end
+        dirlist.each {|d| File.chmod(0755, d)}
+      end
+
+      def pack(sourcedir, destfile)
+        Zlib::GzipWriter.open(destfile) do |writer|
+          Archive::Tar::Minitar.pack(sourcedir, writer)
+        end
+      end
+
+      private
+
+      # Find all the valid files in tarfile.
+      #
+      # Symlinks are not supported in modules and a warning is issued
+      #
+      # This check was mainly added to ignore 'x' and 'g' flags from the PAX
+      # standard but will also ignore any other non-standard tar flags.
+      # tar format info: http://pic.dhe.ibm.com/infocenter/zos/v1r13/index.jsp?topic=%2Fcom.ibm.zos.r13.bpxa500%2Ftaf.htm
+      # pax format info: http://pic.dhe.ibm.com/infocenter/zos/v1r13/index.jsp?topic=%2Fcom.ibm.zos.r13.bpxa500%2Fpxarchfm.htm
+      def find_valid_files(tarfile)
+        Archive::Tar::Minitar.open(tarfile).collect do |entry|
+          flag = entry.typeflag
+          if flag.nil? || flag =~ /[[:digit:]]/ && 2 == flag.to_i
+            @logger.warning("Symlinks in modules are unsupported. Please investigate symlink #{entry.name}.")
```

Comment:
I found a hitch in this - Puppet uses `logger.warning` while r10k uses `logger.warn` so there's a mismatch in which method we should call. I'm not sure how viable injecting a logger will be if we have discrepancies like this.